libm/src/math/arch/i586.rs

```diff
@@ -1,37 +1,67 @@
 //! Architecture-specific support for x86-32 without SSE2
 
-use super::super::fabs;
-
 /// Use an alternative implementation on x86, because the
 /// main implementation fails with the x87 FPU used by
 /// debian i386, probably due to excess precision issues.
-/// Basic implementation taken from https://github.com/rust-lang/libm/issues/219.
-pub fn ceil(x: f64) -> f64 {
-    if fabs(x).to_bits() < 4503599627370496.0_f64.to_bits() {
-        let truncated = x as i64 as f64;
-        if truncated < x {
-            return truncated + 1.0;
-        } else {
-            return truncated;
-        }
-    } else {
-        return x;
+///
+/// Based on https://github.com/NetBSD/src/blob/trunk/lib/libm/arch/i387/s_ceil.S
+/// (written by J.T. Conklin <jtc@NetBSD.org>).
+pub fn ceil(mut x: f64) -> f64 {
+    // We save and later restore the FPU control word.
+    let mut cw_orig: u16 = 0;
+    unsafe {
+        core::arch::asm!(
+            "fstcw   ({cw_ptr})",      // Save the cw
+            "movw    ({cw_ptr}), %dx", // ...
+            "orw     $0x0800, %dx",    // Set rounding control to 0b10 (+∞),
+            "andw    $0xfbff, %dx",    // preserving other controls
+            "movw    %dx, ({cw_ptr})", // Apply cw
+            "fldcw    ({cw_ptr})",     // ...
```

Comment:
Sorry, yeah, as @quaternic mentioned this needs a second stack slot. As a microopt this could be `let mut cw_stash = [0u16; 2];` and then these two can be `-4({stash_ptr})` so we save a register vs. two different `u16` locals

Reply:
Actually even better, `let mut cw_stash = MaybeUninit::<[u16; 2]>::uninit();` to save the zero init instruction

Reply:
I got this to work

https://rust.godbolt.org/z/44Tjcx3bb

the `-4` stuff didn't work (also, for `u16`, why `-4`?)

Reply:
No idea, took that from the original asm without checking :)